Batteries/Control/AlternativeMonad.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2025 Devon Tuma. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Devon Tuma
+-/
+import Batteries.Control.OptionT
+
+
+/-!
+# Laws for Monads with Failure
+
+Definitions for monads that also have an `Alternative` instance while sharing the underlying
+`Applicative` instance, and a class `LawfulAlternative` for types where the failure and monad
+structures are compatible in a natural way. More specifically they satisfy:
+
+* `failure >>= g = failure`
+* `x <|> failure = x`
+* `failure <|> y = y`
+* `x <|> y <|> z = (x <|> y) <|> z`
+* `f <$> (x <|> y) = (f <$> x <|> f <$> y)`
+
+`Option`/`OptionT` are the most basic examples, but transformers like `StateT` also preserve
+the lawfulness of this on the underlying monad.
+
+The law `do _ ← x; failure = failure` is true for monads like `Option` and `List` that don't
```

Comment:
In the symbolic language above, this is `seqRight_failure : x *> failure = failure`